library/x509_create.c

```diff
@@ -125,81 +125,288 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
     return( cur );
 }
 
-int mbedtls_x509_string_to_names( mbedtls_asn1_named_data **head, const char *name )
+static int x509_dn_check_special_chars( const char **cur, const char *end )
 {
     int ret = 0;
-    const char *s = name, *c = s;
-    const char *end = s + strlen( s );
-    const char *oid = NULL;
-    const x509_attr_descriptor_t* attr_descr = NULL;
-    int in_tag = 1;
+
+    /* Check for valid escaped characters */
+    if( *cur == end || *( *cur ) != ',' )
+    {
+        ret = MBEDTLS_ERR_X509_INVALID_NAME;
+    }
+
+    return( ret );
+}
+
+static int x509_dn_parse_tag( const char **cur, const char *end,
+                              const x509_attr_descriptor_t **attr_descr )
+{
+    int ret = 0;
+    const char *start = *cur;
+
+    /*
+     * As defined by RFC 1779, the DN is constructed from
+     * <string> | <key> <optional-space> "=" <optional-space> <string>
+     * where optional space is
+     * ( <CR> ) *( " " )
+     *
+     * When parsing the tag:
+     * 1. Skip whitespaces.
+     * 2. Read the characters, to evaluate as a valid tag.
+     * 3. Skip whitespaces until reaching "=" character.
+     */
+
+    /*
+     * Skip whitespaces
+     */
+    while( *cur < end && *( *cur ) == ' ' )
+    {
+        ( *cur )++;
+        start++;
+    }
+
+    /*
+     * Read the next characters to evaluate as a valid tag.
+     */
+    while( *cur < end && *( *cur ) != '=' && *( *cur ) !=' ' )
+    {
+         ( *cur )++;
+    }
+
+    /*
+     * cur character is either '=' or ' '
+     * or we reached the end.
+     */
+
+    if( *cur == end )
+    {
+        ret = MBEDTLS_ERR_X509_INVALID_NAME ;
+        goto exit;
+    }
+
+    if( ( *attr_descr = x509_attr_descr_from_name
+                        ( start, *cur - start ) ) == NULL )
+    {
+        ret = MBEDTLS_ERR_X509_UNKNOWN_OID;
+        goto exit;
+    }
+
+    /*
+     * Skip all optional whitespaces
+     */
+    while( *cur < end && *( *cur ) == ' ' )
+    {
+        ( *cur )++;
+    }
+
+    /*
+     * The only valid character now is '='
+     */
+    if( *cur == end || *( *cur ) != '=' )
```

Comment:
Question: I have been wondering about this. From the RFC and the comments you have through the code, it looks that a stand-alone attribute is a valid DN. For example, something like "C, D, Z=hello" should be acceptable. Right? However, it looks to me that this is not actually handled by the code. Is this also the case?

Also, i do not think we have test cases for this.

Note: I dont mind if what I pointed out above is not supported, but it would be great if this was documented somewhere in the doxygen documentation of the function (and possibly elsewhere in the code)

(P.S. I do not think this was handled in the previous code either.)

Reply:
But this is not a valid DN, AFAIK. There is always a key and a value

Reply:
mmm, Ok perhaps I am missing something. I was just trying to compose the productions in the grammar at RFC 1779 to parse a single tag. It looks to me like this can be done, but I am not sure its correct or that they are actually used in practice. For example, parsing the DN `atag,btag" could be done using:

```
<name> -> <named-component> ( <spaced-separator> )
<named-component> -> <attribute>
<attribute> -> <string>
<string> -> *( <stringchar> | <pair> )
etc...
```

Reply:
As for tests, there is [this one](https://github.com/ARMmbed/mbedtls/pull/784/files#diff-0807f152469bce0d302ec7282849e58bR156), but perhaps we could add another tests, with the seperator character after

Reply:
Perhaps I am missing something, but the RFC states:
>The keyword specification is optional in the BNF, but mandatory for
   this specification.  This is so that the same BNF may be used for the
   related specification on User Friendly Naming [5].  When this
   specification is followed, the attribute type keywords must always be
   present.

And ( from the BNF grammar)
~~~
<attribute> ::= <string>
           | <key> <optional-space> "=" <optional-space> <string>
~~~

Reply:
Ok